### PR TITLE
Treat CxPlatWorker as a full object

### PR DESCRIPTION
**Pull request by ThadHouse**
## Description

Currently, if you have an app that is creating a datapath outside of msquic, that datapath always shares the same worker context as msquic. There are some cases where this is not the desired behavior, and instead a completely separate worker context should be used, such as if you need to guarantee there is always a free worker, and you have limited the msquic execution context with QUIC_PARAM_GLOBAL_EXECUTION_CONFIG.

This PR changes CxPlatWorker into a full object, with a default instance used. All the datapath init functions take a Worker Manager, and if that manager is null, the default one is used in its place.

## Testing

Alll existing tests will implicitly hit this path, and cover any necessary testing.

## Documentation

No documentation changes, as using CxPlat directly generally isn't a supported scenario.

### CODE REVIEW

**Comment by ThadHouse** [comment]
@microsoft-github-policy-service agree

**Comment by nibanks** [comment]
Generally looks good. Let's just rename WorkerManager to WorkerPool. I think I like that better. Thanks!

File: src/inc/quic_datapath.h [hunk 1]
```diff
@@ -379,6 +379,19 @@ typedef struct CXPLAT_TCP_DATAPATH_CALLBACKS {
 
 } CXPLAT_TCP_DATAPATH_CALLBACKS;
 
+typedef struct CXPLAT_WORKER CXPLAT_WORKER;
```
Comment (nibanks) [inline src/inc/quic_datapath.h:382]: Maybe move this stuff to quic_platform.h, after/around `CXPLAT_EXECUTION_CONTEXT`?

**Review by nibanks — DISMISSED** [review]

**Comment by nibanks** [comment]
>          core.lib(library.obj) : error LNK2001: unresolved external symbol CxPlatWorkerPoolInit [D:\a\msquic\msquic\src\bin\winkernel\msquicpriv.kernel.vcxproj]
>          core.lib(library.obj) : error LNK2001: unresolved external symbol CxPlatWorkerPoolUninit [D:\a\msquic\msquic\src\bin\winkernel\msquicpriv.kernel.vcxproj]

Kernel mode doesn't currently rely on the worker pool.

**Comment by ThadHouse** [comment]
> > ```
> >      core.lib(library.obj) : error LNK2001: unresolved external symbol CxPlatWorkerPoolInit [D:\a\msquic\msquic\src\bin\winkernel\msquicpriv.kernel.vcxproj]
> >      core.lib(library.obj) : error LNK2001: unresolved external symbol CxPlatWorkerPoolUninit [D:\a\msquic\msquic\src\bin\winkernel\msquicpriv.kernel.vcxproj]
> > ```
> 
> Kernel mode doesn't currently rely on the worker pool.

Fixed. Just made them a no-op in kernel mode. Should be fine, any use of CxPlatAddExecutionContext will assert, and then if any functions are used in the data paths they'll get linker errors.

**Review by nibanks — DISMISSED** [review]

**Comment by ThadHouse** [comment]
Ok. Should now correctly be fixed. Forgot how UNREFERENCED_PARAMETER macro worked.

File: src/inc/quic_platform.h [hunk 2]
```diff
@@ -465,14 +495,15 @@ typedef struct CXPLAT_EXECUTION_CONTEXT {
     uint64_t NextTimeUs;
     volatile BOOLEAN Ready;
 
-} CXPLAT_EXECUTION_CONTEXT;
+} CXPLAT_EXECUTION_CONTEXT;
```
Comment (nibanks) [inline src/inc/quic_platform.h:498]: Adding a trailing space? 😮

Comment (ThadHouse) [reply src/inc/quic_platform.h:498]: Nope. It was 2 trailing spaces 🤣

**Comment by nibanks** [comment]
```
"D:\a\msquic\msquic\src\perf\lib\perflib.kernel.vcxproj" (default target) (9) ->
         D:\a\msquic\msquic\src\inc\quic_platform.h(440,22): error C2065: 'WorkerPool': undeclared identifier [D:\a\msquic\msquic\src\perf\lib\perflib.kernel.vcxproj]
         D:\a\msquic\msquic\src\inc\quic_platform.h(440,34): error C2059: syntax error: '(' [D:\a\msquic\msquic\src\perf\lib\perflib.kernel.vcxproj]
         D:\a\msquic\msquic\src\inc\quic_datapath.h(408,14): error C2061: syntax error: identifier 'QUIC_EXECUTION_CONFIG' [D:\a\msquic\msquic\src\perf\lib\perflib.kernel.vcxproj]
         D:\a\msquic\msquic\src\inc\quic_datapath.h(428,10): error C2061: syntax error: identifier 'QUIC_EXECUTION_CONFIG' [D:\a\msquic\msquic\src\perf\lib\perflib.kernel.vcxproj]
         D:\a\msquic\msquic\src\inc\quic_platform.h(440,22): error C2065: 'WorkerPool': undeclared identifier [D:\a\msquic\msquic\src\perf\lib\perflib.kernel.vcxproj]
         D:\a\msquic\msquic\src\inc\quic_platform.h(440,34): error C2059: syntax error: '(' [D:\a\msquic\msquic\src\perf\lib\perflib.kernel.vcxproj]
         D:\a\msquic\msquic\src\inc\quic_platform.h(440,22): error C2065: 'WorkerPool': undeclared identifier [D:\a\msquic\msquic\src\perf\lib\perflib.kernel.vcxproj]
         D:\a\msquic\msquic\src\inc\quic_platform.h(440,22): error C2065: 'WorkerPool': undeclared identifier [D:\a\msquic\msquic\src\perf\lib\perflib.kernel.vcxproj]
         D:\a\msquic\msquic\src\inc\quic_datapath.h(408,14): error C2061: syntax error: identifier 'QUIC_EXECUTION_CONFIG' [D:\a\msquic\msquic\src\perf\lib\perflib.kernel.vcxproj]
         D:\a\msquic\msquic\src\inc\quic_platform.h(440,34): error C2059: syntax error: '(' [D:\a\msquic\msquic\src\perf\lib\perflib.kernel.vcxproj]
         D:\a\msquic\msquic\src\inc\quic_datapath.h(428,10): error C2061: syntax error: identifier 'QUIC_EXECUTION_CONFIG' [D:\a\msquic\msquic\src\perf\lib\perflib.kernel.vcxproj]
         D:\a\msquic\msquic\src\inc\quic_platform.h(440,34): error C2059: syntax error: '(' [D:\a\msquic\msquic\src\perf\lib\perflib.kernel.vcxproj]
         D:\a\msquic\msquic\src\inc\quic_datapath.h(408,14): error C2061: syntax error: identifier 'QUIC_EXECUTION_CONFIG' [D:\a\msquic\msquic\src\perf\lib\perflib.kernel.vcxproj]
         D:\a\msquic\msquic\src\inc\quic_datapath.h(408,14): error C2061: syntax error: identifier 'QUIC_EXECUTION_CONFIG' [D:\a\msquic\msquic\src\perf\lib\perflib.kernel.vcxproj]
         D:\a\msquic\msquic\src\inc\quic_datapath.h(428,10): error C2061: syntax error: identifier 'QUIC_EXECUTION_CONFIG' [D:\a\msquic\msquic\src\perf\lib\perflib.kernel.vcxproj]
         D:\a\msquic\msquic\src\inc\quic_datapath.h(428,10): error C2061: syntax error: identifier 'QUIC_EXECUTION_CONFIG' [D:\a\msquic\msquic\src\perf\lib\perflib.kernel.vcxproj]
         D:\a\msquic\msquic\src\perf\lib\SecNetPerfMain.cpp(230,5): error C2064: term does not evaluate to a function taking 1 arguments [D:\a\msquic\msquic\src\perf\lib\perflib.kernel.vcxproj]
         D:\a\msquic\msquic\src\perf\lib\SecNetPerfMain.cpp(236,14): error C2660: 'CxPlatDataPathInitialize': function does not take 6 arguments [D:\a\msquic\msquic\src\perf\lib\perflib.kernel.vcxproj]
         D:\a\msquic\msquic\src\perf\lib\SecNetPerfMain.cpp(238,9): error C3861: 'CxPlatWorkerPoolUninit': identifier not found [D:\a\msquic\msquic\src\perf\lib\perflib.kernel.vcxproj]
         D:\a\msquic\msquic\src\perf\lib\SecNetPerfMain.cpp(282,9): error C3861: 'CxPlatWorkerPoolUninit': identifier not found [D:\a\msquic\msquic\src\perf\lib\perflib.kernel.vcxproj]
```

**Comment by nibanks** [comment]
Looks like there are lots of test failures.

**Comment by ThadHouse** [comment]
Those are all the same tests that are failing after #4518 was merged

**Comment by nibanks** [comment]
> Some of those are failing on main after #4518 was merged. But there are a few that look to be me.

Crap, I thought everything was passing in that PR.

**Review by nibanks — APPROVED** [review]

**Pull request merged**